### ISSUE TITLE
jsx-handler-names, stateless components and object destructuring

### ISSUE THREAD

**Issue by renato** [jsx-eslint/eslint-plugin-react#1531]
I'm seeing something related to #346 in my eslint errors:

![image](https://user-images.githubusercontent.com/3530/32691110-2dcd18f2-c6ea-11e7-9598-20d38c170604.png)

I solved it by changing the code, but shouldn't it work considering _input_ is just a prop?

**Comment by ljharb** [comment]
Try `const handleChange = input.onChange;` first.

**Comment by renato** [comment]
@ljharb I did solve it by changing the code as I said but the point is: that code shouldn't yield an error, should it?

The same object destructuring works without any trick in normal components.

**Comment by ljharb** [comment]
Ah, fair point. If the same code (with `const { input } = this.props;` and then using `input.onChange`) passes in a non-SFC, then there’s definitely a bug.

The bug might be that the SFC warns, or that the stateful component doesn’t warn, but either way they should behave the same.

**Comment by renato** [comment]
My mistake: the message happens with a stateful component too.

Actually, it happens even if not destructuring the object: `onChange={props.input.onChange}` and `onChange={this.props.input.onChange}`.

In other words, the rule always accepts a `props.onEvent` but never a nested `props.nested.onEvent`, and it's the same when destructuring.

https://github.com/yannickcr/eslint-plugin-react/blob/4f3fc51e32467afbeccc4954765f4312e8e76edf/lib/rules/jsx-handler-names.js#L39

I think it's a bug because if `props.onChange` is acceptable, why `props.input.onChange` wouldn't be?

**Comment by ljharb** [comment]
It shouldn't be acceptable in general because flat props are way better than nested ones :-)

However, I agree that it should be equally acceptable to *this rule*.

**Comment by acherkashin** [comment]
@ljharb Any updates on this issue?

**Comment by ljharb** [comment]
@acherkashin no; but a PR with failing tests would be very helpful.

**Comment by ljharb** [comment]
It seems like this can be covered by setting `eventHandlerPropPrefix` to false. Happy to reopen if not.

**Comment by riteshjagga** [comment]
I don't understand how setting `eventHandlerPropPrefix` to `false` fixes this issue.

`eventHandlerPropPrefix` applies to the prop name and also when a handler is assigned as it is from the function props as in 
```
<MyComponent onChange={props.onChange} />
```

What if I want to code like this
```
const MyInput = ({onChange}) => {
 ...
  return <input type='text' onChange={onChange} />
}
```

**Comment by ljharb** [comment]
It's indeed stylistic, and if you prefer `on` you should configure it that way.

That said, if setting it to `false` stops reporting on those two examples but is reporting on `onChange`, that's a bug - could you please file a new issue?

**Comment by riteshjagga** [comment]
It is not reporting `onChange` either and it seems that it doesn't report at all. So, I'm not sure if a bug should be opened per what you said but you can let me know again after going through this observation.

I tried to inspect code https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L144 to see what is happening:

 - Setting `eventHandlerPropPrefix: false`, sets the `eventHandlerPropPrefix` to `null` - [Line 97](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L97)
 - This then sets the `PROP_EVENT_HANDLER_REGEX` to `null` - [Line 104](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L104)
 - This then sets `propIsEventHandler` to `null` - [Line 140](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L140)
 - And then neither of the if condition executes - [Line 144](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L144) and [Line 157](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/lib/rules/jsx-handler-names.js#L157)

Please let me know what do you think?

**Comment by ljharb** [comment]
Right, i expect them all not to warn if it’s set to false. If you want the prefix “on” you’d need to set it to that.

**Comment by riteshjagga** [comment]
Sorry, I'm confused.

> i expect them all not to warn if it’s set to false>

It seems this is desired

> It seems like this can be covered by setting eventHandlerPropPrefix to false.

This is what you suggested above and closed the bug and I took it as a solution to what is reported.

So both these look contradictory. Am I wrong in understanding the conversation?

---

Also if you can let me know if a new bug needs to open for the [code situation mentioned in this link](https://github.com/jsx-eslint/eslint-plugin-react/issues/1531#issuecomment-1502905549).

**Comment by ljharb** [comment]
Maybe I'm confused :-) are you saying that you *want* it to warn on `onChange={onChange}`, or you want it *not* to warn? Which one are you unable to achieve?

**Comment by riteshjagga** [comment]
I do not want `onChange={onChange}` to be warned because it comes from destructuring the props

but I want onChange={changeHandler} to be warned because it doesn't use `handle` prefix.

**Comment by ljharb** [comment]
The rule definitely doesn't - and isn't intended to - differentiate based on whether it comes from props or not. If the prop is named `onChange` then you can rename it when you destructure it to `handleChange`, for example.

**Comment by riteshjagga** [comment]
OK. Thank you!

**Comment by Bro3Simon** [comment]
I believe this rule doesn't work with destructuring because when you destructure you are creating a local variable.

This produces and error.
``` JavaScript
export function Component({onClick}) {
  return (
    <Button
      onClick={onClick}
    >
      Click Me
    </Button>
  );
}
```

This works fine.
``` JavaScript
export function Component(props) {
  return (
    <Button
      onClick={props.onClick}
    >
      Click Me
    </Button>
  );
}
```
If you destructure you will unfortunately have to rename on destructure in order for it to work.
``` JavaScript
export function Component({onClick: handleClick}) {
  return (
    <Button
      onClick={handleClick}
    >
      Click Me
    </Button>
  );
}
```
Configuration
``` JSON
    "react/jsx-handler-names": [
      "error",
      {
        "checkInlineFunction": true,
        "checkLocalVariables": true
      }
    ],
```

**Comment by ljharb** [comment]
Yes, forcing you to rename on destructure is the intended consequence of the rule here.
